### PR TITLE
[Ready] Respawn functionality

### LINKED ISSUE #48: The player doesn't respawn when dead
The player doesn't respawn when he dies, and this causes funny effects.

I killed myself with a rocket launcher, and it caused me to die and fall down to the ground. Although I couldn't move anymore, I was still able to jump and shoot guns.
openrw

### PR DESCRIPTION
**Pull request by husho**
Respawning is now handled, related opcodes seemed to work.

Closes #48

- [x] Respawning to closest place
- [x] Island overrides for hospitals and jails
- [x] Raw position overrides
- [x] Handling money
- [x] Free health care
- [x] Getting out of jail free
- [x] Penalty calculation based on number of wanted stars
- [x] Fading with correct timing and colours
- [x] Advancing time
- [x] Clearing weapons
- [x] Improve dying state for character object
- [x] Opcodes
- [x] Refactor? fading stuff, so you can fade by calling a function
- [x] Show wasted / busted text
- [x] ~~~Camera flying away~~~ (not going to do now)
- [x] Increment death/jail stat
- [x] Mission override restart fixed

### CODE REVIEW

File: rwengine/src/ai/PlayerController.hpp [hunk 1]
```diff
@@ -33,6 +33,11 @@ class PlayerController : public CharacterController {
 
     void enterNearestVehicle();
 
+	bool isWasted();
+	bool isBusted();
+
+	void restartLogic();
```
Comment (ShFil119) [inline rwengine/src/ai/PlayerController.hpp]: Some comment would be useful. ;)

File: rwengine/src/engine/GameState.hpp [hunk 2]
```diff
@@ -319,8 +319,55 @@ class GameState {
...
+    bool overrideNextRestart;
     glm::vec4 nextRestartLocation;
+    std::vector<glm::vec4> hospitalRestarts, policeRestarts; 
+
+    void addHospitalRestart(const glm::vec4 location) {
+    	hospitalRestarts.push_back(location);
+    }
+
+    void addPoliceRestart(const glm::vec4 location) {	
+		policeRestarts.push_back(location);
```
Comment (ShFil119) [inline rwengine/src/engine/GameState.hpp]: To reformat.

File: rwengine/src/engine/GameState.hpp [hunk 3]
```diff
@@ -319,8 +319,55 @@ class GameState {
     /** Objects created by the current mission */
     std::vector<GameObject*> missionObjects;
 
-    bool overrideNextStart;
+    bool overrideNextRestart;
```
Comment (ShFil119) [inline rwengine/src/engine/GameState.hpp]: It should be initialized.

File: rwengine/src/objects/CharacterObject.cpp [hunk 4]
```diff
@@ -645,3 +645,9 @@ void CharacterObject::useItem(bool active, bool primary) {
         }
     }
 }
+
+void CharacterObject::clearInventory() {
+	for (int slot = 0; slot < kMaxInventorySlots; ++slot) {
```
Comment (ShFil119) [inline rwengine/src/objects/CharacterObject.cpp]: Range loop?

File: rwengine/src/ai/PlayerController.hpp [hunk 5]
```diff
@@ -33,6 +33,11 @@ class PlayerController : public CharacterController {
 
     void enterNearestVehicle();
 
+	bool isWasted();
+	bool isBusted();
```
Comment (ShFil119) [inline rwengine/src/ai/PlayerController.hpp]: Maybe todo comment?

File: rwengine/src/objects/CharacterObject.cpp (hunk 4 above)
Comment (husho) [reply rwengine/src/objects/CharacterObject.cpp]: in this case range loop make things complicated, because we are getting a value which we need to unpack

Comment (ShFil119) [reply rwengine/src/objects/CharacterObject.cpp]: You are right, I haven't noticed usage of slot.

File: rwengine/src/engine/GameState.hpp [hunk 6]
```diff
@@ -323,8 +325,65 @@ class GameState {
...
+        float closest = 10000.f;
+        glm::vec4 result;
+
+        ZoneData* playerZone = world->data->findZoneAt(playerPosition);
+
+        std::vector<glm::vec4> iter;
+        int islandOverride;
+
+        if (type == Hospital) {
+            iter = hospitalRestarts;
```
Comment (danhedron) [inline rwengine/src/engine/GameState.hpp]: You could avoid copying the list here

File: rwengine/src/script/modules/GTA3ModuleImpl.inl [hunk 7]
```diff
@@ -4032,11 +4032,9 @@ bool opcode_016b(const ScriptArguments& args) {
...
     @arg angle Angle
 */
-void opcode_016c(const ScriptArguments& args, ScriptVec3 coord, const ScriptFloat angle) {
-    RW_UNIMPLEMENTED_OPCODE(0x016c);
-    RW_UNUSED(coord);
-    RW_UNUSED(angle);
-    RW_UNUSED(args);
+void opcode_016c(const ScriptArguments& args, ScriptVec3 coord, const ScriptFloat heading) {
+	coord = script::getGround(args, coord);
+    args.getState()->addHospitalRestart(glm::vec4(coord, heading));
```
Comment (danhedron) [inline rwengine/src/script/modules/GTA3ModuleImpl.inl]: Mixing tabs and spaces?

**Comment by mythrillex** [comment]
when fail the mission the game dont restart after this merge, looks like gtasa but in gta3 the game restart right ?

File: rwengine/src/script/modules/GTA3ModuleImpl.inl (hunk 7 above)
Comment (husho) [reply rwengine/src/script/modules/GTA3ModuleImpl.inl]: yes, I can fix these exact lines, but whole file should be reformatted anyway

File: rwengine/src/engine/GameState.hpp (hunk 6 above)
Comment (husho) [reply rwengine/src/engine/GameState.hpp]: Will do

**Comment by mythrillex** [comment]
hmm sorry, if u fail the first mission, the game will restart the mission, but with this respawn the game dont occur

**Comment by husho** [comment]
#443 patch required to make missions failable upon death

**Comment by husho** [comment]
Ready for review

**Comment by mythrillex** [comment]
if u kill 8-ball the game make a respawn loop, restarting the mission over and over, its cause the game dont restart things when start the mission again (example if u break a lampost, when restart the lampost still broken)
i make this mistake of kill 8-ball when the mission restart, before he enter in car, I kept the mouse pressed and killed him with the gun :v
image

**Comment by husho** [comment]
Yes, you are not supposed to shoot or move or anything during cutscenes, also when cutscene starts it should clear all projetables and make player safe. But that's not related to this PR. Maybe I'll do it later.

**Pull request merged**